File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.py
n = int(input())
input_list=list(map(int,input().split()))

max_val=0
index_list=[]
def max_value(arr:list,n):
    global max_val
    if len(arr)==n:
        # 최댓값 계산함수 호출
        temp=swap_cal(arr,n)
        if temp>max_val:
            max_val=temp
    else:
        for i in range(n):
            if i not in index_list:
                index_list.append(i)
                max_value(index_list,n)
                index_list.pop()


# 인덱스 교환하면서 값을 계산해주는 함수
def swap_cal(arr:list,n)->int:
    max_val=0
    for i in range(len(arr)-1):
        max_val+=abs(input_list[arr[i]]-input_list[arr[i+1]])

    return max_val


max_value(index_list,n)

print(max_val)

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Main{
    //입력
//    6
//20 1 15 8 4 10
    static int maxV=0;
    static List<Integer> inputList;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        List<Integer> list = new ArrayList<>();
        for(int i=0;i<N;i++){
            list.add(sc.nextInt());
        }
        inputList=list;
        recur(new ArrayList<>(),N);
        System.out.println(maxV);
    }

    private static void recur(List<Integer> input, int N){
        if(input.size() == N){
            // 여기서 맥스값 비교 갱신 하고
               int temp =0;
               for(int i =0;i<N-1;i++){
                   temp+=Math.abs(inputList.get(input.get(i))-inputList.get(input.get(i+1)));
               }
               if(temp>maxV) maxV=temp;

               return;
        }
        // 인덱스 값이아닌 오브젝트 값이 나오자나 그걸 리스트에 넣어 그렇다면 ... 인덱스를 찾는게 아니넹 ?
        // input 에는 인덱스 값이 담겨야 할것같ㅇ느딩 ..
        for (int i = 0; i < N; i++) {
            if(!input.contains(i)){
                input.add(i);
                recur(input,N);
                input.remove((Integer) i);
            }
        }
    }
}
